zigen-project/zen-remote
# create and destroy rendering unit

## Context

**Rendering Unit** is a set of resources to render a single framebuffer, to which texture, vertex buffer, shader, and other parameters are associated.
zen share the list of rendering units and the resources associated to the rendering units with the remote client (ex., Oculus Quest). The remote client perform rendering by traversing the rendering units.

We use gRPC to share rendering units and associated resource over network. It's unacceptable to perform network stuff in the main thread as it's too slow. So I implement job and job queue. Jobs pushed in a job queue are performed in a worker thread managed by the job queue. Jobs are performed in the order they are pushed.

Resources shared between client and server (including rendering units) have the same id (uint64). Both use this id to specify the resource. This resource ID is monotonically increasing, but since it is uint64, I do not think it is a problem.

## Summary

- [x] Define protobuf for rendering unit.
- [x] Create job and job queue.
- [x] Create simple gprc_server (client side)
- [x] Send request to create and destroy rendering unit

In this system,

Client (Oculus Quest) does
- UDP client for peer discovery
- TCP server for peer discovery
- gRPC server for main communication.

Server (zen) does,
- UDP server for peer discovery
- TCP client for peer discovery
- gRPC client for main communication

## How to check behavior

- Build and run using following changes
  - zen: Use the branch of this pull request
  - zen-oculus-display-system: https://github.com/zigen-project/zen-oculus-display-system/pull/20
  - zen: https://github.com/zigen-project/zen/pull/190
- Start immersive display system.
- See log and confirm that the connection established
- Plug and unplug pointing device, you will see the following log in Oculus Quest.

```log
10-15 19:55:50.495 23419 23826 D ZEN[Remote]: New rendering unit: 0
10-15 19:55:50.545 23419 23800 I ZEN[Remote]: Server Found: 172.16.0.36
10-15 19:56:01.018 23419 23826 D ZEN[Remote]: Delete rendering unit: 0
10-15 19:56:05.980 23419 23826 D ZEN[Remote]: New rendering unit: 1
10-15 19:56:07.704 23419 23914 D ZEN[Remote]: Delete rendering unit: 1
10-15 19:56:10.346 23419 23826 D ZEN[Remote]: New rendering unit: 2
10-15 19:56:14.339 23419 23914 D ZEN[Remote]: Delete rendering unit: 2
10-15 19:56:25.776 23419 23826 D ZEN[Remote]: New rendering unit: 3
10-15 19:56:28.177 23419 23826 D ZEN[Remote]: Delete rendering unit: 3
10-15 19:56:31.029 23419 23914 D ZEN[Remote]: New rendering unit: 4
10-15 19:56:34.753 23419 23826 D ZEN[Remote]: Delete rendering unit: 4
```

## What's left

#11 #12
